code/model1.py

```diff
@@ -1,9 +1,91 @@
+import torch
+import torch.nn as nn
+import torch.autograd as autograd
+import torch.optim as optim
+import utils
+import numpy as np
 
 
-class Model1(Model):
+class Model1(nn.Module, Model):
+    """
+    Contains the code for Model1.
+    Architecture - sr_embedding -> linear_layer -> relu_activation -> linear_layer -> predicted_email_emb -> loss
+     - sr_embedding is calculated by concatenation of sender and receiver embeddings
+     - Loss is calculated as L2 loss between predicted_email_embedding and email_representation. The email
+     representation is obtained by averaging embeddings from pre-trained word2vec model
+    """
+
+    def __init__(self, epochs=10):
+        self.epochs = epochs
+        super(Model1, self).__init__()
+        # embedding lookup for 150 users each have 50 dimension representation
+        self.embedding_layer = nn.Embedding(150, 50)
+        # first hidden layer, linear layer with weights 200x500
+        # this should be the size of <sender+receiver representation>
+        self.h1_layer = nn.Linear(100, 500)
+        # ReLU activation used
+        self.relu = nn.ReLU()
+        # final linear layer that outputs the predicted email representation
+        self.email_layer = nn.Linear(500, 50)
+
+    def forward(self, s_id, r_id):
+        """
+        Input are 2 autgrad variables representing sender_id and receiver_id. These ids will be used to lookup the
+        embeddings. Does a forward pass and returns the predicted email representation
+        :param s_id:
+        :param r_id
+        :return: email representation
+        """
+        # extract the sender and receiver embedding
+        s_emb = self.embedding_layer(s_id)
+        r_emb = self.embedding_layer(r_id)
+        # simple concatenation of sender and receiver embedding
+        sr_emb = torch.cat((s_emb, r_emb), 1)
+        h1 = self.relu(self.h1_layer(sr_emb))
+        email_reps = self.email_layer(h1)
+        return email_reps
+
+    def get_average_rep(self, word_reps):
+        """
+        Assumes that word_reps is a numpy 2d array with every row as vector representation of word.
+        Calculates the mean across all rows to get average email representation
+        :param word_reps:
+        :return:
+        """
+        return np.mean(word_reps, axis=0)
+
     def train(self, emails, w2v):
-        pass
+        loss_criteria = nn.MSELoss()
+        optimizer = optim.RMSprop(self.parameters(), lr=0.001, alpha=0.6, momentum=0.6)
+
+        for epoch in range(self.epochs):
+            epoch_loss = 0.0
+            for i in range(len(emails)):
+                sender_id = utils.get_userid(emails[i, 0])
+                email_word_reps = w2v.get_sentence(emails[i, 2])
+                # if no word_rep was found for any of the words in the emails, ignore this case
+                if len(email_word_reps) == 0:
+                    continue
+                email_rep = self.get_average_rep(email_word_reps)
```

Comment:
Would it be better to just call it inline?
If we reuse it for model3's receiver embeddings, then I think the function is justified.

Reply:
done